lib/very_good_infinite_list.dart

```diff
@@ -1,468 +1,322 @@
-  late _Debouncer _debouncer;
+class _InfiniteListState extends State<InfiniteList> {
+  late final CallbackDebouncer _debounce;
 
-  void _onListStateChanged() {
-    final state = _controller.value;
-    if (_isFailure(state)) {
-      widget.onError?.call(context, _controller.fetch, state.exception.value);
-    }
-  }
-
-  bool _isFailure(_ListState<T> state) {
-    return state.status.isFailure && state.exception != _ListException.none;
-  }
+  ScrollController? _scrollController;
 
   @override
   void initState() {
     super.initState();
-    _controller = _ListController<T>(widget.itemLoader);
-    _debouncer = _Debouncer(delay: widget.debounceDuration);
-    _scrollController = widget._scrollController ?? ScrollController()
-      ..addListener(_onScroll);
-    _controller
-      ..addListener(_onListStateChanged)
-      ..fetch();
+    _debounce = CallbackDebouncer(widget.debounceDuration);
+    _initScrollController();
+    WidgetsBinding.instance?.addPostFrameCallback((_) {
+      _attemptFetch();
+    });
   }
 
   @override
-  void dispose() {
-    _debouncer.dispose();
-    _controller
-      ..removeListener(_onListStateChanged)
-      ..dispose();
-    _scrollController.removeListener(_onScroll);
-    if (widget._scrollController == null) {
-      _scrollController.dispose();
+  void didUpdateWidget(InfiniteList oldWidget) {
+    super.didUpdateWidget(oldWidget);
+
+    if (widget.scrollController != oldWidget.scrollController) {
+      _initScrollController();
     }
-    super.dispose();
-  }
 
-  @override
-  void didUpdateWidget(covariant InfiniteList<T> oldWidget) {
-    super.didUpdateWidget(oldWidget);
-    if (widget._scrollController != oldWidget._scrollController) {
-      _scrollController.removeListener(_onScroll);
-      _scrollController = widget._scrollController ?? ScrollController()
-        ..addListener(_onScroll);
+    if (widget.itemCount != oldWidget.itemCount ||
+        widget.hasReachedMax != oldWidget.hasReachedMax) {
+      WidgetsBinding.instance?.addPostFrameCallback((_) {
+        _attemptFetch();
+      });
     }
   }
 
   @override
-  Widget build(BuildContext context) {
-    return ValueListenableBuilder(
-      valueListenable: _controller,
-      builder: (context, _ListState<T> state, child) {
-        final itemCount = state.hasReachedMax == false
-            ? state.items.length + 1
-            : state.items.length;
-
-        if (state.status == _ListStatus.loading) {
-          return widget.builder.loading(context);
-        }
-
-        if (state.items.isEmpty) {
-          return state.exception != _ListException.none
-              ? widget.builder.error(
-                  context,
-                  _controller.fetch,
-                  state.exception.value,
-                )
-              : widget.builder.empty(context);
-        }
+  void dispose() {
+    _scrollController?.removeListener(_attemptFetch);
+    _scrollController?.dispose();
```

Comment:
```suggestion
    if (widget.scrollController == null) {
      _scrollController?.dispose();
    }
```
You should not dispose the controller that has been provided from outside